All in one commit, forced to hard rebase

vaavaa opened the pull request.

Commit:
All in one commit, forced to hard rebase

vaavaa commented:
Вроде по чек листу сверился.

gennady-bars commented on src/utils/config.js:
```diff
@@ -6,6 +6,8 @@ export const sauceLimit = -485;
 //URL на API
 export const API_URL = "https://norma.nomoreparties.space/api/ingredients";
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу и использовать везде в коде проекта, добавляя в конце изменяющуюся часть

gennady-bars commented on src/utils/config.js:
```diff
@@ -6,6 +6,8 @@ export const sauceLimit = -485;
 //URL на API
 export const API_URL = "https://norma.nomoreparties.space/api/ingredients";
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу и использовать везде в коде проекта, добавляя в конце изменяющуюся часть


`apiURL ` и есть эта часть универсальная.

vaavaa replied:
Да, это у меня от прошлого спринта осталось. Подправлю.

vaavaa replied:
для верной сборки лучше использовать **npm install --legacy-peer-deps**, (машина самовольно притащила новые библиотеки, и целостность связей нарушилась)

APi паз подправил.

gennady-bars reviewed (commented):
Работу придется отклонить без потери итерации:

* Не запустилось приложение . Скрин ошибки http://joxi.ru/L21BQEbczN51qA 

Еще обратите внимание:

* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу и использовать везде в коде проекта, добавляя в конце изменяющуюся часть

Commit:
package.json fixed

Commit:
1) npm install --legacy-peer-deps
2) clear api paths

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,23 +1,58 @@
+    const history = useHistory();
+    const location = useLocation();
+    //Любой ввод пути, в строке браузера : переход по ссылке
+    const background = history.action === 'PUSH' && location.state && location.state.background;
```
отличная логика отображения попапа или страницы

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,81 @@
+        <div className={style.container}>
+            <div className={style.login_container}>
+                <h3 className="text text_type_main-medium mb-6">Вход</h3>
+                <form onSubmit={handleFormSubmit} className="form">
```
отлично, что везде сабмит вешается на форму, а не клик на кнопку

Многие тут ошибаются

gennady-bars commented on src/components/burger-ingredients/ingredient-details/ingredient-details.jsx:
```diff
@@ -1,33 +1,54 @@
+
+    useEffect(() => {
+        if (ingredients.length <= 0) {
+            dispatch(getIngredientsFromServer())
```
` dispatch(getIngredientsFromServer());` нужно получать ингредиенты с сервера только  в `App`, чтобы любые страницы проекта имели их уже для отрисовки.

gennady-bars reviewed (commented):
Здравствуйте. Работа проделана огромная, но есть некоторые недочеты:

* восстановление пароля не работает. Не идет переход на путь `/reset-password"` после нажатия на восстановить
* Если открыть попап ингредиента и перезагрузить страницу, приложение ломается с ошибкой. Скрин http://joxi.ru/bmoKL1Vu71MLpm
* Лучше сделать цвет ссылок по фигме. Скрин http://joxi.ru/ZrJKJ9yuby1dWm . А то слишком темный цвет на черном фоне
* ` dispatch(getIngredientsFromServer());` нужно получать ингредиенты с сервера только  в `App`, чтобы любые страницы проекта имели их уже для отрисовки.



**Можно лучше**

*  при клике на логотип обычно идет переход на главную страницу
* Лучше ограничивать кол-во срабатываний фильтраций массивов при помощи хука `useMemo`.  Это касается таких операций, как `filter`, `map`, `forEach`, `reduce` и тд . Подробнее https://ru.reactjs.org/docs/hooks-reference.html#usememo 
* Модальное окно с деталями ингредиента лучше поместить в `App`. Тогда будет легче сделать логику при перезагрузке страницы

Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Commit:
1) dispatch(getIngredientsFromServer()) in App - done
2) цвет ссылок по фигме - done
3) click логотип переход на главную страницу - done
4) попап ингредиента и перезагрузить страницу  ошибка - in progress

Commit:
1) dispatch(getIngredientsFromServer()) in App - done
2) цвет ссылок по фигме - done
3) click логотип переход на главную страницу - done
4) попап ингредиента и перезагрузить страницу  ошибка - ошибку убрал, но попап пропал. (День коту под хвост.)
5) useMemo - done
6) Восстановление пароля -done.
7) Модальное окно в App - done

Commit:
Ingredient opened in modal by link and path, finally.

vaavaa commented:
В итоге получилось реализовать все пункты.

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* по заданию нужно сделать логику, при которой бы показывался попап с деталями ингредиента, если кликнуть по ингредиенту, но если при открытом попапе перезагрузить страницу, то должна уже показываться страница с этими деталями ингредиента. Тоже самое должно быть, если скопировать ссылку на ингредиент при открытом попапе и вставить в окно браузера. (сейчас уже разрешается показывать страницу при вставке ссылки в браузер, а при перезагрузке страницы можно открытый попап). Но если скопировать ссылку и вставить в браузер - должна быть уже страница с деталями ингредиента

Commit:
ingredients open in separated page, no more popup by path

Commit:
ingredients open in separated page if popup 've been closed.

vaavaa commented:
прошу прощения, видно я долблюсь уже в шары и не понимаю как верно нужно. Перечитал задание. Изменил, и не уверен что лучшим способом.

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -60,27 +62,36 @@ const App = () => {
                 <main className={style.content_body}>
                     <Switch location={background || location}>
                         <Route path="/" exact={true}>
+                            {deletePopupCookie}
```
` {deletePopupCookie}` нужно удалить. Так нельзя делать. Из-за этого ошибки в консоли.

gennady-bars reviewed (commented):
Все еще нужно кое-что исправить:

* Вам нужно использовать средства роутинга для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://reactrouter.com/web/example/modal-gallery . В ней используются `location` и `background`, которые передаются в `Switch`.  Пожалуйста, посоветуйтесь с наставником.
* В этой ссылке показано, как отображать попап, если нет `background`. Присмотритесь к коду там
* В консоли ошибки. Скрин http://joxi.ru/1A5YdBpSzxNB6A
* ` {deletePopupCookie}` нужно удалить. Так нельзя делать. Из-за этого ошибки в консоли.
* Так не и отображается страница с деталями, если вставить ссылку в браузер

Commit:
error cleared, modal window stays save

vaavaa commented:
Победил, в итоге. Думал уже окончательно отупел... но нет еще сопративляюсь.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

К следующему спринту нужно будет подправить:

* нужно будет выровнять страницу с деталями по середине. Скрин http://joxi.ru/12MRE9jtw9e0dA

```js

 const background = history.action === 'PUSH' && location.state && location.state.background;

```

Вот это было бы достаточно. Раньше как раз так и нужно было, чтобы при перезагрузке страницы показывалась бы страница (а не открытый попап). Так что это очень хороший вариант и не нужно `useEffect` никакие

Commit:
Merge remote-tracking branch 'origin/main' into sprint-3/step-1

# Conflicts:
#	package.json
#	src/components/app-header/app-header.js
#	src/components/app/app.jsx
#	src/components/burger-ingredients/burger-ingredients.jsx
#	src/components/burger-ingredients/ingredient-card/ingredient-card.jsx
#	src/components/burger-ingredients/ingredient-details/ingredient-details.jsx
#	src/components/profile/menu-profile/menu-profile.jsx
#	src/components/profile/menu-profile/menu-profile.module.css
#	src/index.tsx
#	src/pages/reset-password/reset-password.jsx
#	src/pages/reset-password/reset-password.module.css
#	src/pages/restore-password/restore-password.jsx
#	src/services/actions/user.jsx
#	src/services/reducers/user.jsx
#	src/utils/config.js
#	src/utils/utils.js
#	yarn.lock
